test/bats/provider.bats

```diff
@@ -0,0 +1,164 @@
+
+  run kubectl get secret --namespace=test foosecret
+  [ "$status" -eq 0 ]
+
+  result=$(kubectl --namespace=test get secret foosecret -o jsonpath="{.metadata.labels.environment}")
+  [[ "${result//$'\r'}" == "test" ]]
+
+  result=$(kubectl --namespace=test get secret foosecret -o jsonpath="{.metadata.labels.secrets-store\.csi\.k8s\.io/managed}")
+  [[ "${result//$'\r'}" == "true" ]]
+
+  # There isn't really an event we can wait for to ensure this has happened.
+  for i in {0..60}; do
+      result="$(kubectl --namespace=test get secret foosecret -o json | jq '.metadata.ownerReferences | length')"
+      if [[ "$result" -eq 2 ]]; then
+          break
+      fi
+      sleep 1
+  done
+  [[ "$result" -eq 2 ]]
+
+  # Wait for secret deletion in a background process.
+  kubectl --namespace=test wait --for=delete secret foosecret &
+  WAIT_PID=$!
+
+  # Trigger deletion implicitly by deleting only owners.
+  kubectl --namespace=test delete -f $CONFIGS/nginx-env-var.yaml
+  echo "Waiting for foosecret to get deleted"
+  wait $WAIT_PID
+
+  # Ensure it actually got deleted.
+  run kubectl --namespace=test get secret foosecret
+  [ "$status" -eq 1 ]
+}
+
+@test "3 SecretProviderClass in different namespace not usable" {
+  kubectl create namespace negative-test-ns
+  kubectl --namespace=negative-test-ns apply -f $CONFIGS/nginx-env-var.yaml
+  kubectl --namespace=negative-test-ns wait --for=condition=PodScheduled --timeout=60s pod -l app=nginx
+  POD=$(kubectl get pod -l app=nginx -n negative-test-ns -o jsonpath="{.items[0].metadata.name}")
+
+  wait_for_success "kubectl describe pod $POD -n negative-test-ns | grep 'FailedMount.*failed to get secretproviderclass negative-test-ns/vault-foo-sync.*not found'"
+}
+
+@test "4 Pod with multiple SecretProviderClasses" {
+  POD=nginx-multiple-volumes
+  kubectl --namespace=test apply -f $CONFIGS/nginx-multiple-volumes.yaml
+  kubectl --namespace=test wait --for=condition=Ready --timeout=60s pod $POD
+
+  result=$(kubectl --namespace=test exec $POD -- cat /mnt/secrets-store-1/bar1)
+  [[ "$result" == "hello-sync1" ]]
+  result=$(kubectl --namespace=test exec $POD -- cat /mnt/secrets-store-2/bar2)
+  [[ "$result" == "hello-sync2" ]]
+
+  result=$(kubectl --namespace=test get secret foosecret-1 -o jsonpath="{.data.username}" | base64 -d)
+  [[ "$result" == "hello-sync1" ]]
+  result=$(kubectl --namespace=test get secret foosecret-2 -o jsonpath="{.data.pwd}" | base64 -d)
+  [[ "$result" == "hello-sync2" ]]
+
+  result=$(kubectl --namespace=test exec $POD -- printenv | grep SECRET_1_USERNAME | awk -F"=" '{ print $2 }' | tr -d '\r\n')
+  [[ "$result" == "hello-sync1" ]]
+  result=$(kubectl --namespace=test exec $POD -- printenv | grep SECRET_2_PWD | awk -F"=" '{ print $2 }' | tr -d '\r\n')
+  [[ "$result" == "hello-sync2" ]]
+}
+
+wait_for_success() {
+  echo $1
+  for i in {0..60}; do
+      if eval "$1"; then
+          return
+      fi
+      sleep 1
+  done
+  # Fail the test.
+  [ 1 -eq 2 ]
+}
```

Comment:
I wonder if it'd be useful to put functions like this in a _helpers.bash, then load them into bats with `load _helpers`.

Reply:
👍 done